RocksDB add ConditionExpression and wire into put()

stevelorddremio opened the pull request.

kylep-dremio committed:
Separate generic store tests to tiered-tests.

kylep-dremio committed:
Add support for RocksDB.

kylep-dremio committed:
Add a test for getRefs and fix RocksDBStore imple.

kylep-dremio committed:
Clean up rocks tests to properly reset state.
Fix an ordering issue in AbstractTestStore#getRefs.

kylep-dremio committed:
Change SerDe to use Protobuf instead of Kryo.
Fix an incorrect file path as well.

Upgrade the version of Guava as it seems that 29.0 has a bug
in Multimap#index where it will non-deterministically have issues
reducing the keys and throw duplicate key errors, which shows up
in RocksDBStore#save. I could not reproduce this in 30.1.

kylep-dremio committed:
Remove extraneous MongoDB changes.

kylep-dremio committed:
Change to use oneof pattern in proto3.

Note that map and repeated aren't allowed in oneof,
so wrap them in another message to work around this.

kylep-dremio committed:
Address code review comments.

kylep-dremio committed:
Add missing deserialization tests.

kylep-dremio committed:
Address code review comments.

kylep-dremio committed:
Address code review comments.

stevelorddremio committed:
Backing stores, fixed checkstyle issues, override tests for methods not implemented in RocksDB and MongoDB for this baseline.

stevelorddremio committed:
Revised key to store SaveOps Multimap in RocksDBStore.save() method to avoid ColumnFamilyHandles. Use ValueType directly now.

stevelorddremio committed:
RocksDB putIfAbsent implemented

kylep-dremio committed:
Merge branch 'main_origin' into create-rocks-store

# Conflicts:
#	pom.xml
#	versioned/pom.xml
#	versioned/tiered-tests/src/main/java/com/dremio/nessie/versioned/impl/SampleEntities.java
#	versioned/tiered-tests/src/main/java/com/dremio/nessie/versioned/tests/AbstractTestStore.java
#	versioned/tiered/mongodb/pom.xml
#	versioned/tiered/mongodb/src/test/java/com/dremio/nessie/versioned/store/mongodb/TestMongoDBStore.java

kylep-dremio committed:
Partial commit.

stevelorddremio committed:
RocksDB ConditionExpression evaluator

stevelorddremio committed:
Merge branch 'main' into rocks-queries

stevelorddremio committed:
Build out RocksDB ConditionExpression evaluator

stevelorddremio committed:
RocksDB add RocksL1 consumer for ConditionExpression consumption.

stevelorddremio committed:
RocksDB Added implementations of L1, Branch and Tag for ConditionExpression evaluation.

stevelorddremio committed:
RocksDB ConditionExpression tests

stevelorddremio committed:
RockDB ConditionExpression tests L1

stevelorddremio committed:
RockDB checkstyle corrections

stevelorddremio committed:
RocksDB add condition expression tests

stevelorddremio committed:
RocksDB checkstyle updates

stevelorddremio committed:
RocksDB ConditionExpression tests

stevelorddremio committed:
RocksDB L1 Condition evaluation tests

stevelorddremio committed:
RocksDB L1 ConditionExpression tests

stevelorddremio committed:
RocksDB ConditionExpression L1 tests

stevelorddremio committed:
RocksDB ConditionExpression L1 tests

stevelorddremio committed:
RocksDB Move common methods up to RocksBaseValue

stevelorddremio committed:
RocksDB ConditionExpression for Ref

stevelorddremio committed:
RocksDB Add ID ConditionExpression for L1 and Ref

Commit:
Implement RocksL2 and RocksL3

stevelorddremio committed:
RocksDB checkstyle

stevelorddremio committed:
Rocks Add further Ref tests

stevelorddremio committed:
Rocks checkstyle

tifflhl committed:
Implement RockCommitMetadata

- TODO: Complete unit tests

tifflhl committed:
Address review comments and fix checkstyle warnings

kylep-dremio committed:
Merge branch 'rocks-queries-l2-l3' into create-rocks-store

# Conflicts:
#	versioned/tiered/rocksdb/pom.xml
#	versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBStore.java

kylep-dremio committed:
Complete rebase of RocksDB with producer/consumer pattern.

stevelorddremio committed:
Rocks add Fragment ConditionExpression evaluation

stevelorddremio committed:
Rocks ConditionExpression Fragment tests

stevelorddremio committed:
Rocks Branch Commits Size ConditionExpression evaluator

stevelorddremio committed:
RocksDB Add condition expression handling to RocksDBStore. Add tests to AbstractTestStore

kylep-dremio committed:
Update some of the protobuf formats for RocksDB.

kylep-dremio commented on pom.xml:
```diff
@@ -139,7 +139,9 @@
     <prometheus.version>0.9.0</prometheus.version>
+    <rocksdb.version>6.13.3</rocksdb.version>
     <reactor.version>2020.0.3</reactor.version>
+    <rocksdb.version>6.13.3</rocksdb.version>
```
Duplicate.

stevelorddremio replied:
Removed

kylep-dremio commented on versioned/tiered/dynamodb/src/test/java/com/dremio/nessie/versioned/impl/condition/TestExpressions.java:
```diff
@@ -28,7 +28,6 @@
 
   private final Entity av0 = Entity.ofBoolean(true);
   private final Entity av1 = Entity.ofBoolean(false);
-  private final Entity av2 = Entity.ofString("mystr");
```
Why changes to dynamo?

stevelorddremio replied:
We removed IF_NOT_EXISTS from ExpressionFunction.FunctionName as per Jacques comment on Dec 21
"Jacques Nadeau 1:58 PM
It looks like you're referring to the static SetClause.ifNotExists method. If we're not using it anywhere, yes, makes sense to remove. (edited) "

That broke the DynamoDB test which I fixed.
This was added as part of https://github.com/projectnessie/nessie/pull/741
This PR is based on #741 so includes the same change.

kylep-dremio commented on versioned/tiered/mongodb/pom.xml:
```diff
@@ -35,12 +35,6 @@
       <artifactId>nessie-versioned-tiered-impl</artifactId>
       <version>${project.version}</version>
     </dependency>
-    <dependency>
```
Why changes to Mongo?

stevelorddremio replied:
This looks like your change on 2021-01-18  4e5b3b7f7248fac79219e3dd6b2005ad24620f45

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Condition.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * A set of conditions that are asserted against an Entity.
+ */
+public class Condition {
```
Shouldn't have public classes unless there's a reason to do so.

stevelorddremio replied:
Removed public to make it package local.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Condition.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class Condition {
+
+  public List<Function> functionList;
```
Should never have public members - or as Laurent already suggested we can use Immutables.

stevelorddremio replied:
I updated the member to make this private then added a getting to return an immutable copy.

kylep-dremio replied:
Can you look at using Immutables (https://immutables.github.io/) instead?

stevelorddremio replied:
Will do

stevelorddremio replied:
Done

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Evaluator.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.List;
+import java.util.regex.Pattern;
+
+public interface Evaluator {
```
What is this for?

stevelorddremio replied:
Removed public

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * An expression that is asserted against an Entity.
+ */
+public class Function {
```
Public....

stevelorddremio replied:
Now package local

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,65 @@
+  public static final String EQUALS = "equals";
+  public static final String SIZE = "size";
+
+  final String operator;
```
Why not private?

stevelorddremio replied:
Is is now.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,65 @@
+ * An expression that is asserted against an Entity.
+ */
+public class Function {
+  public static final String EQUALS = "equals";
```
Why not using ExpressionFunction.FunctionName ?

kylepbit replied:
Was this missed?

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,65 @@
+   * @param ex object to compare
+   * @return true if this is equal to provided object
+   */
+  public boolean equals(Function ex) {
```
This is odd - if we've added an equals method it should be an override for the Object#equals method, which means we also need to override  Object#hashCode. This on it's own is going to cause issues.

stevelorddremio replied:
Fixed.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,65 @@
+
+  /**
+   * Compares for equality with a provided Function object.
+   * @param ex object to compare
```
Why is this called ex?

stevelorddremio replied:
ex -> expression. Not that obvious on reflection. Will change to `function`.

kylep-dremio replied:
ex typically refers to exception by convention, fyi.

stevelorddremio replied:
ack

stevelorddremio replied:
Done

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,65 @@
+        && this.value.equals(ex.value));
+  }
+
+  public String getOperator() {
```
Why public?

stevelorddremio replied:
This was bouncing between packages when I first wrote this. Now package local.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,192 @@
+   * @return the List Entity.
+   */
+  Entity toEntity(Stream<Id> idStream) {
+    List<Id> idList = idStream.collect(Collectors.toList());
```
Why not:
return Entity.ofList(idStream.map(Id::toEntity).collect(Collectors.toList()));

stevelorddremio replied:
That's better. I had trouble visualizing this as a pipeline to make it work. Makes sense now.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,192 @@
+   * @return the List Entity.
+   */
+  Entity toEntity(Stream<Id> idStream, int position) {
+    return idStream.collect(Collectors.toList()).get(position).toEntity();
```
Why not:
return idStream.skip(position).findFirst().orElseThrow(NoSuchElementException::new).toEntity();

This removes the need to create a list - impl needs to be verified.

stevelorddremio replied:
Impl works nicely. I guess this is more efficient because it removed the need to convert to a list first.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,192 @@
+   */
+  boolean evaluateStream(Function function, Stream<Id> stream) {
+    // EQUALS will either compare a specified position or the whole stream as a List.
+    final List<String> path = Evaluator.splitPath(function.getPath());
```
Seems like we take the path and convert to string on creation of the Function, then break apart after the fact when evaluating. Why not keep as separate parts until needed?

stevelorddremio replied:
This is refactored now to store ExpressionPath in Function. I then drill down into that as needed in the evaluate method.
I'm reluctant to leach more ExpressionPath/ConditionExpression objects into this design so I can socialize it as a replacement for those DynamoDB specific implementations.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,192 @@
+ */
+abstract class RocksBaseValue<C extends BaseValue<C>> implements BaseValue<C> {
+
+  public static final String ID = "id";
```
Why public?

stevelorddremio replied:
Now package local.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,192 @@
+        // Remove enclosing brackets.
+        final List<String> arguments = Arrays.asList(segment.split("[()]"));
+        if (arguments.size() == 1) { // compare complete list
+          return (toEntity(stream).equals(function.getValue()));
```
Nit: Extra brackets around the expression.

stevelorddremio replied:
Removed and below too.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,192 @@
+      if (function.getOperator().equals(Function.EQUALS)) {
+        // Remove enclosing brackets.
+        final List<String> arguments = Arrays.asList(segment.split("[()]"));
+        if (arguments.size() == 1) { // compare complete list
```
I don't understand the logic for the split of 1 arg vs. 2 for equals, can this be commented?

stevelorddremio replied:
Added comment describing behaviour.

stevelorddremio replied:
I refactored this using ExpressionPath. I hope it is more obvious.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,192 @@
+          return (toEntity(stream, position).equals(function.getValue()));
+        }
+      } else if (function.getOperator().equals(Function.SIZE)) {
+        return (stream.count() == (int)function.getValue().getNumber());
```
Why casting down to int instead of up to long?

stevelorddremio replied:
Agreed. Will keep as long.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,56 @@
+      final String segment = path.get(0);
+      if (segment.equals(ID)) {
+        result &= ((path.size() == 1)
+          && (function.getOperator().equals(Function.EQUALS))
```
Why the extra braces?

stevelorddremio replied:
Removed, also for other evaluate() implementations.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,56 @@
+          && (function.getOperator().equals(Function.EQUALS))
+          && (getId().toEntity().equals(function.getValue())));
+      } else if (segment.equals(VALUE)) {
+        result &= ((path.size() == 1)
```
If this returns false, seems like it should stop the loop at that point in which case &= would not be relevant.

stevelorddremio replied:
Agreed. Updated for all evaluate() methods to result if false;

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBConditionVisitor.java:
```diff
@@ -0,0 +1,170 @@
+  */
+  @Override
+  public Condition visit(final ConditionExpression conditionExpression) {
+    //    final String functions = conditionExpression.getFunctions().stream()
```
Why is dead code still here?

stevelorddremio replied:
Spotted that and removed already.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBConditionVisitor.java:
```diff
@@ -0,0 +1,170 @@
+    //      .collect(Collectors.joining("& "));
+
+    Condition holder = new Condition();
+    holder.functionList = conditionExpression.getFunctions().stream()
```
Have Condition take the list as part of construction to simplify the code here.

stevelorddremio replied:
Agreed. Done.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBConditionVisitor.java:
```diff
@@ -0,0 +1,170 @@
+        case EQUALS:
+          // Special case SIZE, as the object representation is not contained in one level of ExpressionFunction.
+          if (isSize(arguments.get(0))) {
+            return new Function(Function.SIZE,
```
Why are we converting children into strings? We then end up string parsing during evaluation, I think this needs to be re-evaluated, we're currently using string as an intermediary representation instead of using a proper object representation.

stevelorddremio replied:
This is refactored now to store ExpressionPath in Function. I then drill down into that as needed in the evaluate method.
I'm reluctant to leach more ExpressionPath/ConditionExpression objects into this design so I can socialize it as a replacement for those DynamoDB specific implementations.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBPathVisitor.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Visitor for creating String path representations from ExpressionPaths.
+ */
+class RocksDBPathVisitor implements ExpressionPath.PathVisitor<Boolean, String, RuntimeException> {
```
This seems like a common pattern to need to convert to String - should it be moved into ExpressionPath as a default visitor?

stevelorddremio replied:
This is refactored now to store ExpressionPath in Function. I then drill down into that as needed in the evaluate method.
This visitor class is no longer needed.

stevelorddremio committed:
RocksDB review updates

stevelorddremio reviewed (commented):
Still comments to address.

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBStore.java:
```diff
@@ -0,0 +1,359 @@
+  public <C extends BaseValue<C>> void put(SaveOp<C> saveOp, Optional<ConditionExpression> condition) {
+    if (condition.isPresent()) {
+      RocksBaseValue consumer = RocksSerDe.getConsumer(saveOp);
+      loadSingle(saveOp.getType(), saveOp.getId(), (C) consumer);
```
How is this supposed to work in multi-threaded environments? That was a key consideration for this as we discussed before.

stevelorddremio replied:
This will be made transactional.

stevelorddremio replied:
This is the final comment left to implement. Doing that today (Feb 8th)

stevelorddremio replied:
Done.

stevelorddremio committed:
Merge branch 'create-rocks-store' into rocks-fragment-query

stevelorddremio committed:
Rocks refactor Function to use ExpressionPath instead of String to store path

stevelorddremio committed:
RocksDB clean up unused visitor class and methods

stevelorddremio committed:
Rocks checkstyle

stevelorddremio committed:
RocksDB clean up unused classes

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Condition.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * A set of conditions that are asserted against an Entity.
+ */
+class Condition {
```
I had asked about https://immutables.github.io/ before but don't see any changes on that, or response....

stevelorddremio replied:
Done

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Evaluator.java:
```diff
@@ -0,0 +1,41 @@
+   * @param path the input string path.
+   * @return the split separate parts of the path.
+   */
+  static List<String> splitPath(String path) {
```
Is this still necessary?

stevelorddremio replied:
No removed.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * An expression that is asserted against an Entity.
+ */
+class Function {
```
Why not use immutables: https://immutables.github.io/ ?

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,181 @@
+  boolean evaluateStream(Function function, Stream<Id> stream) {
+    // EQUALS will either compare a specified position or the whole stream as a List.
+    if (function.getOperator().equals(Function.EQUALS)) {
+      if (function.getPath().getRoot().getChild().isPresent()
```
Can this be cleaned up at all using orElseGet() or something on the Optional?

stevelorddremio replied:
I kind of simplified I guess.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,64 @@
+  public boolean evaluate(Condition condition) {
+    for (Function function: condition.getFunctionList()) {
+      if (function.getPath().getRoot().isName()) {
+        ExpressionPath.NameSegment nameSegment = function.getPath().getRoot().asName();
```
Please use final where applicable.

stevelorddremio replied:
Changed and for the other classes.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,64 @@
+
+        switch (segment) {
+          case ID:
+            if (!(!nameSegment.getChild().isPresent()
```
Please carry the outside ! through the condition.

kylepbit replied:
Why does it matter if nameSegment child is present? It's not used here?

stevelorddremio replied:
It was an extra level of validation to make sure the condition wasn't `id.something` instead of just `id`

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBStoreConfig.java:
```diff
@@ -0,0 +1,52 @@
+public abstract class RocksDBStoreConfig {
+
+  @Default
+  public String getRefTableName() {
```
Can you change this to use the prefix as the mongo store does so it's all matched up now?

stevelorddremio replied:
I have now. I assume you want me to change the other getters here too? In which case what family column does getTreeTableName relate to? Also where is this class actually used? It didn't show up in finds.

kylepbit replied:
Please see how Mongo does it: https://github.com/stevelorddremio/nessie/blob/main/versioned/tiered/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDBStore.java#L145

stevelorddremio replied:
I was a bit lost about mapping ColumnFamilyHandles to ValueTypes. I updated the map to use the same pattern as MongoDBStore and updated MongoDBStoreConfig with the extra TablePrefix atttribute, but I don't think this is quite correct.
Also the Config object is an Immutable and it doesn't look like we build it correctly in the TestRocksDBStore. So probably need a chat in the morning to clear this up. I set this up in rocks-config branch instead.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,64 @@
+          case ID:
+            if (!(!nameSegment.getChild().isPresent()
+                && function.getOperator().equals(Function.EQUALS)
+                && getId().toEntity().equals(function.getValue()))) {
```
The ID check seems to be duplicated across values objects, can this be deduplicated please.

stevelorddremio replied:
Yes, done.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksFragment.java:
```diff
@@ -0,0 +1,129 @@
+   * @return an entity list of keys
+   */
+  private Entity keysAsEntityList(Stream<Key> keys) {
+    List<Key> keyListIn = keys.collect(Collectors.toList());
```
Why not:

return Entity.ofList(keys.map(k -> Entity.ofList(k.getElements().stream().map(Entity::ofString))));

stevelorddremio replied:
Thanks. Updated.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksL1.java:
```diff
@@ -0,0 +1,208 @@
+          case INCREMENTAL_KEY_LIST:
+            if (nameSegment.getChild().isPresent() && function.getOperator().equals(Function.EQUALS)) {
+              if (nameSegment.getChild().get().asName().getName().equals(CHECKPOINT_ID)) {
+                result &= checkpointId.toEntity().equals(function.getValue());
```
Still old code here using &=

stevelorddremio replied:
Oops. All replaced now. Did a search.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksL3.java:
```diff
@@ -0,0 +1,96 @@
+          // TODO: ConditionExpression is currently not supported for TREE.
+          // TODO: Implement a case for TREE if serialization of a Stream of KeyDelta to Entity is supported.
+          // Invalid Condition Function.
+          return false;
```
Shouldn't this throw an UnsupportedException?

stevelorddremio replied:
Good point but I'm going to remove this TODO. I think evaluate() should behave the same for all tests that don't pass by returning false. Otherwise we should throw an exception, for instance, when a NameSegment in the ConditionException does not match an attribute in the object on which we run the evaluate().

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksRef.java:
```diff
@@ -0,0 +1,336 @@
+  @Override
+  public boolean evaluate(Condition condition) {
+    // Retrieve entity at function.path
+    if (type == RefType.BRANCH) {
```
Why type here, but on line 58 it's this.type?

stevelorddremio replied:
Hmm. Inconsistency. I removed the this on line 58.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksRef.java:
```diff
@@ -0,0 +1,336 @@
+            }
+            break;
+          case TYPE:
+            if (!(!nameSegment.getChild().isPresent()
```
This structure shows up all over, which indicates it should be refactored to reduce duplication.

stevelorddremio replied:
Moved common code to RocksBaseValue and refactored each class.

stevelorddremio committed:
RocksDB PR updates

tifflhl committed:
RocksDB Store - Implement Delete

tifflhl committed:
Complete implementation for Rocks store - delete()

- TODO: Unit and integration tests

stevelorddremio committed:
RocksDB PR updates

stevelorddremio committed:
RocksDB override Function.equals() and hashCode()

kylepbit commented on versioned/tiered/rocksdb/pom.xml:
```diff
@@ -0,0 +1,122 @@
+      <plugin>
+        <groupId>org.xolstice.maven.plugins</groupId>
+        <artifactId>protobuf-maven-plugin</artifactId>
+        <version>0.6.1</version>
```
Is this plugin used elsewhere/should this version be in dependencyManagement?

stevelorddremio replied:
Yes, it is used by nessie-quarkus. I added it to the nessie parent pom in the PluginManagement section and properties then referenced the plugin from each child pom.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,71 @@
+   * @param function object to compare
+   * @return true if this is equal to provided object
+   */
+  boolean equals(Function function) {
```
I had left a comment before about this vs. Object#equals that has not been addressed.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,196 @@
+  boolean evaluateStream(Function function, Stream<Id> stream) {
+    // EQUALS will either compare a specified position or the whole stream as a List.
+    if (function.isEquals()) {
+      ExpressionPath.PathSegment pathSegment = function.getPath().getRoot().getChild().orElse(null);
```
Final?

stevelorddremio replied:
Yes, added.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,196 @@
+   * @param function the condition function to check it is for equality
+   * @return true if both nameSegment is childless and function has an equality operator
+   */
+  boolean nameSegmentChildlessAndEquals(ExpressionPath.NameSegment nameSegment, Function function) {
```
static?

stevelorddremio replied:
Yes. Changed.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,66 @@
+            }
+            break;
+          case VALUE:
+            if (!(nameSegmentChildlessAndEquals(nameSegment, function)
```
Please don't keep ! outside of complex boolean operations, please carry through using DeMorgans.

stevelorddremio replied:
Ok, updated the logic.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksL1.java:
```diff
@@ -0,0 +1,214 @@
+  @Override
+  public boolean evaluate(Condition condition) {
+    boolean result = true;
+    for (Function function: condition.getFunctions()) {
```
It seems like this structure is common across all the evaluate() method impls, how can you refactor to reduce code duplication?

stevelorddremio replied:
Moved loop implementation up into RocksBaseValue and added abstract method to be implemented by child classes.

stevelorddremio committed:
RocksDB PR updates

stevelorddremio committed:
RocksDB PR updates

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -30,13 +32,32 @@
+      return false;
+    }
+
+    Function function = (Function) object;
```
Final - please ensure this is done in the future as we've discussed before.

stevelorddremio replied:
Sorry, updated. Haven't taken a break

stevelorddremio committed:
Rocks PR update

tifflhl committed:
Add integration tests for delete()

stevelorddremio committed:
RocksDB move evaluate loop up to base class

stevelorddremio committed:
RocksDB remove implements Evaluator from derived classes of RocksBaseValue

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Evaluator.java:
```diff
@@ -0,0 +1,38 @@
+   * @param function the condition to check
+   * @return true if the condition is met
+   */
+  boolean evaluateSegment(ExpressionPath.NameSegment nameSegment, Function function);
```
It seems odd this is on the interface - I think likely what you want is an abstract method on the BaseValue. When would I call evaluateSegment on it's own?

stevelorddremio replied:
I tried to keep evaluator methods together but I guess this is more implementation detail so moved to RocksBaseValue

kylepbit replied:
Adding to the interface means that all children need to implement this, however it's a detail of the Rocks implementation. "Together" is a separate concept from "must implement and is public" which is what the interface is saying.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,225 @@
+        return toEntity(stream).equals(function.getValue());
+      } else { // compare complete list
+        if (pathSegment.isPosition()) { // compare individual element of list
+          int position = pathSegment.asPosition().getPosition();
```
Final...

stevelorddremio replied:
Doh. Ok, updated.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksBaseValue.java:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  @Override
+  public abstract boolean evaluateSegment(ExpressionPath.NameSegment nameSegment, Function function);
```
As mentioned earlier, this shouldn't be on the interface as it's only called from evaluate() - why would it need to be on the interface?

stevelorddremio replied:
Agreed. Updated as suggested and renamed it to evaluateFunction. Also reduced dependencies by removing NameSegment from parameters. It comes from Function anyway, so no need to pass separately.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,64 @@
+      case ID:
+        // ID is considered a leaf attribute, ie no children. Ensure this is the case
+        // in the ExpressionPath.
+        if (!idEvaluates(nameSegment, function)) {
```
Why not `return idEvaluates(nameSegment, function);` ?

stevelorddremio replied:
Now the loop has gone that makes sense again. I update all evaluateFunction() methods.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksCommitMetadata.java:
```diff
@@ -0,0 +1,64 @@
+      case VALUE:
+        // VALUE is considered a leaf attribute, ie no children. Ensure this is the case
+        // in the ExpressionPath.
+        if (!nameSegmentChildlessAndEquals(nameSegment, function)
```
Why not:
`return nameSegmentChildlessAndEquals(...) && byteValue.toStringUtf8()...);` ?

For the secondary part, why do you have to convert to a string first?

stevelorddremio replied:
No we don't need this intermediate step. Will keep at binary.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksFragment.java:
```diff
@@ -0,0 +1,114 @@
+        // Invalid Condition Function.
+        return false;
+    }
+    return true;
```
I'd rather that each case return directly instead of falling through to here, it should simplify the if blocks in many cases.

stevelorddremio replied:
Agreed. Refactored here and other evaluateFunction() implementations.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksL1.java:
```diff
@@ -0,0 +1,206 @@
+            if (!Entity.ofNumber(distanceFromCheckpoint).equals(function.getValue())) {
+              return false;
+            }
+          } else {
```
If you eliminated the else block here and at line 136, you could replace the break with a return false (provided the other if/else if blocks return in the true/false cases directly).

stevelorddremio replied:
I refactored this before I read this. I think I have captured your simplification sentiment and it looks cleaner. Let me know if not.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksL3.java:
```diff
@@ -0,0 +1,90 @@
+  @Override
+  public boolean evaluateSegment(ExpressionPath.NameSegment nameSegment, Function function) {
+    final String segment = nameSegment.getName();
+    if (segment.equals(ID)) {
```
This could be simplified as such (in the same pattern as suggested above):

final String segment = nameSegment.getName();
return segment.equals(ID) ? idEvaluates(nameSegment, function) : false;

stevelorddremio replied:
Agreed. Makes sense. Done.

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksRef.java:
```diff
@@ -0,0 +1,321 @@
+    } else if (type == RefType.TAG) {
+      return evaluateTag(nameSegment, function);
+    }
+    return true;
```
Should this return true in this case, or false?

stevelorddremio replied:
Good catch. Now returns false.

kylepbit commented on versioned/tiered/tiered-impl/src/main/java/com/dremio/nessie/versioned/impl/IdMap.java:
```diff
@@ -40,7 +40,7 @@
  * of their original state so one can see what items changed over time.
  */
-class IdMap implements Iterable<Id> {
+public class IdMap implements Iterable<Id> {
```
Why was this made public?

stevelorddremio replied:
I required it at some point but it is no longer needed. I have reverted this change now.

stevelorddremio committed:
RocksDB reduce parameters passed & PR comments

stevelorddremio committed:
RocksDB checkstyle

stevelorddremio committed:
Rocks PR comments

tifflhl committed:
Fix failing unit/integration tests

stevelorddremio committed:
RocksDB PR comments

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/Function.java:
```diff
@@ -0,0 +1,106 @@
+   */
+  @Override
+  public boolean equals(Object object) {
+    if (object == null) {
```
You don't need the null check, instanceof will return false of object is null.

stevelorddremio replied:
Ok, done.

kylepbit commented on versioned/tiered/rocksdb/src/test/java/com/dremio/nessie/versioned/store/rocksdb/TestConditionExpressions.java:
```diff
@@ -0,0 +1,563 @@
+
+    final Condition expectedCondition = ImmutableCondition.builder()
+        .addFunctions(ImmutableFunction.builder()
+        .operator(Function.EQUALS)
```
These should be indented at least one more to indicate they're not part of the ImmutableCondition. Should be applied everywhere this comment is applciable.

stevelorddremio replied:
Checkstyle complained last time but was ok this time. Done

kylepbit commented on versioned/tiered/rocksdb/src/test/java/com/dremio/nessie/versioned/store/rocksdb/TestConditionExpressions.java:
```diff
@@ -0,0 +1,563 @@
+    final String path = createPath();
+    final ConditionExpression ex = ConditionExpression.of(ExpressionFunction.equals(ofPath(path), MAP_ENTITY));
+
+    final Condition expectedCondition = ImmutableCondition.builder()
```
The creation of the expectedCondition seems to be duplicated with the exception of the entity that is placed, please refactor to reduce duplication.

stevelorddremio replied:
Refactored

kylepbit commented on versioned/tiered/rocksdb/src/test/java/com/dremio/nessie/versioned/store/rocksdb/TestConditionExpressions.java:
```diff
@@ -0,0 +1,563 @@
+        .build();
+    equals(expectedCondition, ex);
+  }
+  // TODO: map with list
```
Should this be handled?

stevelorddremio replied:
I don't think so, I asked the community what complexity should be implemented. I only got a response back from Robert who said the AbstractITTieredVersionStore tests should pass. So when we put this together I will add tests if it exposes any errors.

stevelorddremio committed:
Merge remote-tracking branch 'origin/rocks-delete' into rocks-fragment-query

stevelorddremio committed:
Rocks add transactional to put()

kylepbit commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBStore.java:
```diff
@@ -0,0 +1,392 @@
+
+        transaction.delete(columnFamilyHandle, value);
+        transaction.commit();
+        return true;
```
When would this return false?

stevelorddremio replied:
If I understand correctly, the method would return false if the condition expression was not met but the id was found.

kylepbit commented on versioned/tiered/rocksdb/src/test/java/com/dremio/nessie/versioned/store/rocksdb/TestConditionExecutor.java:
```diff
@@ -0,0 +1,810 @@
+    final String path = createPath();
+    final Condition condition = ImmutableCondition.builder()
+        .addFunctions(ImmutableFunction.builder()
+        .operator(Function.EQUALS)
```
This should be indented to indicate not part of ImmutableCondition, applies to this and the following.
Secondly, this is mostly boilerplate, can you refactor to reduce duplication across the tests?

stevelorddremio replied:
This has been done.

kylepbit commented on versioned/tiered/rocksdb/src/test/java/com/dremio/nessie/versioned/store/rocksdb/TestConditionExecutor.java:
```diff
@@ -0,0 +1,810 @@
+    return (RocksL1) new RocksL1()
+        .id(Id.EMPTY)
+        .commitMetadataId(ID)
+        .children(IntStream.range(0, RocksL1.SIZE).mapToObj(x -> ID))
```
Would this be better as: Stream.generate(() -> ID).limit(RocksL1.SIZE) ?

stevelorddremio replied:
It looks more succinct. Done and elsewhere in this class

stevelorddremio committed:
RocksDB final variables

tifflhl committed:
Address code review comments

- Added more delete() tests.
- Modified RocksDBStore - delete() to return false instead of throwing ConditionFailedException, to match the base Store API's Javadoc.
- Modified RocksDBStore - delete() to delete with the id.

stevelorddremio committed:
RocksDB instantiate RocksDBStore from config object. Use getTablePrefix for valueTypeToColumnFamily initializeation.

stevelorddremio committed:
RocksDB checkstyle change

stevelorddremio committed:
Merge remote-tracking branch 'origin/rocks-delete' into rocks-fragment-query

kylep-dremio commented on versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBStore.java:
```diff
@@ -0,0 +1,392 @@
+
+    try (final Transaction transaction = transactionDB.beginTransaction(new WriteOptions(), new OptimisticTransactionOptions())) {
+      final byte[] value = transaction.getForUpdate(new ReadOptions(), columnFamilyHandle, id.toBytes(), true);
+      if (value != null) {
```
Can this be changed to match put() - so that it's value == null and the throw is early?

stevelorddremio replied:
Yes, done

tifflhl committed:
Throw NotFoundException when a matching value of the give key is not returned

stevelorddremio committed:
RockDB test duplication and PR comments

stevelorddremio committed:
RocksDB remove duplicated test code

stevelorddremio committed:
Merge remote-tracking branch 'origin/rocks-delete' into rocks-fragment-query

stevelorddremio committed:
Rocks config Entity to ColumnFamily mapping - almost

stevelorddremio committed:
Rocks config correction

stevelorddremio committed:
Rocks PR update

stevelorddremio committed:
Rocks checkstyle

stevelorddremio committed:
RocksDB added filter to find ValueType by TableName in ValueType to ColumnFamily mapping.

stevelorddremio committed:
RocksDB ColumnFamilies now named using ValueType.getValueName()

stevelorddremio committed:
RocksDB combine createConfig methods.

stevelorddremio committed:
Merge pull request #12 from stevelorddremio/rocks-config

RocksDB config

kylep-dremio committed:
Change transactions to be pessimistic.

kylep-dremio committed:
Add TierdVersionStore and VersionStore tests.

stevelorddremio committed:
Rocks merge main into branch

stevelorddremio committed:
Rocks fixed casting issue in TestConditionExecutor tests

stevelorddremio committed:
Rocks formatting corrections

kylep-dremio committed:
Correct SerDe issue with RocksRef after merge.

stevelorddremio commented:
Replaced by rocks-condition-expr-rebase branch
